File: swift/Sources/AetherNetProtocol/Transport/TransportManager.swift
// SPDX-License-Identifier: MIT

import Foundation

/// Multi-transport manager: routes an outbound send through the best available transport and
/// surfaces every inbound delivery through a single callback tagged with the carrying transport's
/// name. Swift counterpart of the C# `AetherNet.Transport.Services.TransportManager`, the Go
/// `transport.Manager`, and the Python `TransportManager`, reduced to the real selection path the
/// mesh needs:
///
/// - transports are held sorted **ascending** by ``TransportService/powerCostRelative``, so the
///   cheapest is tried first and an expensive last-resort transport (the circuit relay, cost 90)
///   is only reached after every cheaper one has declined; and
/// - ``sendAsync(peerUhid:data:cancellationToken:)`` falls through the ordered transports until one
///   returns `true` or all decline.
///
/// This is exactly the C# manager's "step 6: additional transports, sorted by `PowerCostRelative`
/// (ascending), fall through until one succeeds". Typed BLE / Wi-Fi Direct / NearLink slots are not
/// modelled here — on this SDK every transport (including those) is registered as an ordinary
/// transport and ordered purely by power cost, which is what makes the relay a genuine
/// auto-selected fallback rather than a hand-wired special case.
///
/// On the receive side the manager subscribes to each transport that adopts
/// ``DataReceivingTransport`` (every real mesh transport does) and re-raises its deliveries through
/// ``onDataReceived(_:)`` tagged `(sender, data, via)`, where `via` is the delivering transport's
/// ``TransportService/name`` — the Swift equivalent of the C# manager subscribing to each
/// `ITransportService.DataReceived` event and the Go manager's `dataReceiver` type-assertion.
public final class TransportManager: @unchecked Sendable {

    /// `(senderUhid, data, viaTransportName)` — the manager's received-data contract.
    public typealias DataReceivedHandler = (String, Data, String) -> Void

    private let transports: [any TransportService]
    private let lock = NSLock()
    private var onData: DataReceivedHandler?

    /// Builds a manager over `transports`, ordered ascending by ``TransportService/powerCostRelative``
    /// so the lowest-cost transport is preferred and the highest-cost (e.g. the circuit relay at 90)
    /// is the last-resort fallback. Subscribes to each transport's receive surface (those adopting
    /// ``DataReceivingTransport``); inbound data is re-raised through ``onDataReceived(_:)`` tagged
    /// with that transport's ``TransportService/name``.
    ///
    /// The ordering is stable for equal costs (registration order is preserved), matching the C#
    /// `OrderBy(t => t.PowerCostRelative)` and Go `sort.SliceStable` stable sorts.
    public init(_ transports: [any TransportService]) {
        // Stable ascending sort by power cost (Swift's `sorted(by:)` is not guaranteed stable, so
        // pair each transport with its original index and break ties on it to preserve order).
        self.transports = transports
            .enumerated()
            .sorted { lhs, rhs in
                if lhs.element.powerCostRelative != rhs.element.powerCostRelative {
                    return lhs.element.powerCostRelative < rhs.element.powerCostRelative
                }
                return lhs.offset < rhs.offset
            }
            .map { $0.element }

        for t in self.transports {
            guard let rx = t as? DataReceivingTransport else { continue }
            let via = t.name
            rx.onDataReceived { [weak self] sender, data in
                guard let self = self else { return }
                self.lock.lock(); let cb = self.onData; self.lock.unlock()
                cb?(sender, data, via)
            }
        }
    }

    /// Convenience initializer for the common single-transport case (e.g. the gap-2 acceptance
    /// test wires each node's relay as its only transport).
    public convenience init(_ transports: any TransportService...) {
        self.init(transports)
    }

    /// Registers the callback invoked when any transport delivers data to this node. Arguments are
    /// `(senderUhid, payload, name of the transport that carried it)` — the "via" tag proves which
    /// transport the manager selected on the receive side, mirroring the C#
    /// `TransportManager.DataReceived (sender, data, transportName)` event.
    public func onDataReceived(_ handler: @escaping DataReceivedHandler) {
        lock.lock(); onData = handler; lock.unlock()
    }

    /// Sends `data` to `peerUhid`, trying each available transport in ascending power-cost order
    /// until one succeeds. Returns `true` on the first transport that reports delivery; `false` if
    /// every transport is unavailable or declines. A transport that declines (e.g. the relay with
    /// "no route yet") is skipped and the manager moves to the next candidate — identical to the
    /// C# / Go / Python fall-through.
    @discardableResult
    public func sendAsync(peerUhid: String, data: Data, cancellationToken: CancellationToken? = nil) async -> Bool {
        for t in transports {
            if !t.isAvailable { continue }
            if await t.sendAsync(peerUhid: peerUhid, data: data, cancellationToken: cancellationToken) {
                return true
            }
        }
        return false
    }

    /// Sends a whole buffer to `peerUhid` via the first stream-capable transport that succeeds,
    /// in ascending power-cost order.
    @discardableResult
    public func sendStreamAsync(peerUhid: String, data: Data, cancellationToken: CancellationToken? = nil) async -> Bool {
        for t in transports {
            if !t.isAvailable { continue }
            if await t.sendStreamAsync(peerUhid: peerUhid, data: data, cancellationToken: cancellationToken) {
                return true
            }
        }
        return false
    }

    /// The manager's transports in the order they are tried (ascending power cost). The returned
    /// array is a copy; mutating it does not affect the manager.
    public var orderedTransports: [any TransportService] { transports }
}

File: typescript/src/transport/TransportManager.ts
/**
 * Multi-transport manager that routes each send through the best available transport and,
 * failing that, falls through the remaining transports until one succeeds.
 *
 * Selection order mirrors the C# reference (AetherNet.Transport.Services.TransportManager):
 * typed fast-paths (NearLink → small-payload BLE → Wi-Fi Direct → CircleLink → large-payload BLE)
 * are tried first, then **additional transports sorted by ascending `powerCostRelative`**. The
 * native circuit-relay-v2 transport advertises cost 90 (just below the HTTP relay's 100), so it is
 * auto-selected as the LAST-RESORT serverless fallback — chosen only after every cheaper direct
 * transport has declined, and never hand-wired by the caller.
 *
 * The TypeScript {@link ITransportService} contract is leaner than the C#/Go one (no separate
 * IBle/IWifiDirect/INearLink/ICircleLink marker interfaces, and receive is a single
 * `onDataReceived` callback rather than a multicast event). This manager therefore takes every
 * transport through the `additionalTransports` list, orders them by power cost, and chains each
 * transport's receive callback so delivered data surfaces via {@link TransportManager.onDataReceived}
 * tagged with the transport's `name` — proving which transport the manager selected.
 *
 * SPDX-License-Identifier: MIT
 */

import { ITransportService } from "./ITransportService.js";

/** Aggregate diagnostics for sends routed through the manager. */
export interface TransportManagerMetrics {
  /** Successful sends per transport, keyed by transport name. */
  sendCountByTransport: Map<string, number>;
  /** Bytes sent per transport, keyed by transport name. */
  bytesSentByTransport: Map<string, number>;
  /** Sends that no available transport could satisfy. */
  totalFailures: number;
}

/**
 * Routes sends across a set of transports and surfaces their inbound data through one callback.
 * Construct with the transports this node runs; the manager orders them by ascending power cost so
 * cheap direct links are preferred and the circuit relay (cost 90) is the automatic fallback.
 */
export class TransportManager {
  private readonly transports: ITransportService[];
  private readonly sendCount = new Map<string, number>();
  private readonly bytesSent = new Map<string, number>();
  private totalFailures = 0;
  private disposed = false;

  /** Fired when data arrives via any managed transport: (sender, data, viaTransportName). */
  onDataReceived?: (senderUhid: string, data: Uint8Array, via: string) => void;

  /**
   * @param additionalTransports The transports this node runs. Ordered internally by ascending
   *   `powerCostRelative`, so the cheapest usable transport is tried first and the relay last.
   */
  constructor(additionalTransports: ITransportService[] = []) {
    // Stable sort by ascending power cost (relay's 90 sorts after cheaper direct transports).
    this.transports = [...additionalTransports].sort(
      (a, b) => a.powerCostRelative - b.powerCostRelative,
    );
    this.subscribeToDataEvents();
  }

  /**
   * Sends `data` to `peerUhid`, trying each available transport in ascending-power-cost order until
   * one succeeds. Returns false only if every transport declined.
   */
  async sendAsync(
    peerUhid: string,
    data: Uint8Array,
    cancellationToken?: AbortSignal,
  ): Promise<boolean> {
    for (const transport of this.transports) {
      if (!transport.isAvailable) continue;
      if (await transport.sendAsync(peerUhid, data, cancellationToken)) {
        this.bump(this.sendCount, transport.name, 1);
        this.bump(this.bytesSent, transport.name, data.length);
        return true;
      }
    }
    this.totalFailures += 1;
    return false;
  }

  /**
   * Sends a stream to `peerUhid`, trying each available transport in ascending-power-cost order.
   */
  async sendStreamAsync(
    peerUhid: string,
    stream: ReadableStream<Uint8Array>,
    cancellationToken?: AbortSignal,
  ): Promise<boolean> {
    for (const transport of this.transports) {
      if (!transport.isAvailable) continue;
      if (await transport.sendStreamAsync(peerUhid, stream, cancellationToken)) {
        this.bump(this.sendCount, transport.name, 1);
        return true;
      }
    }
    this.totalFailures += 1;
    return false;
  }

  /** True if any managed transport reports an active connection to `peerUhid`. */
  isConnected(peerUhid: string): boolean {
    return this.transports.some((t) => t.isConnected(peerUhid));
  }

  /** Snapshot of per-transport send counters and total failures. */
  getMetrics(): TransportManagerMetrics {
    return {
      sendCountByTransport: new Map(this.sendCount),
      bytesSentByTransport: new Map(this.bytesSent),
      totalFailures: this.totalFailures,
    };
  }

  /** Detaches from every managed transport's receive callback. */
  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    for (const t of this.transports) t.onDataReceived = undefined;
    this.onDataReceived = undefined;
  }

  // ── internals ──────────────────────────────────────────────────────────────────

  private subscribeToDataEvents(): void {
    for (const transport of this.transports) {
      // Preserve any handler already attached to the transport, then tag with its name.
      const prior = transport.onDataReceived;
      transport.onDataReceived = (sender, data) => {
        prior?.(sender, data);
        this.onDataReceived?.(sender, data, transport.name);
      };
    }
  }

  private bump(map: Map<string, number>, key: string, by: number): void {
    map.set(key, (map.get(key) ?? 0) + by);
  }
}
